server/src/main/java/io/crate/execution/engine/indexing/ShardingUpsertExecutor.java

```diff
@@ -101,8 +102,11 @@ public class ShardingUpsertExecutor
     private volatile boolean createPartitionsRequestOngoing = false;
     private final Predicate<UpsertResults> earlyTerminationCondition;
     private final Function<UpsertResults, Throwable> earlyTerminationExceptionGenerator;
+    private final Runnable validatorsCacheCleaner;
 
     ShardingUpsertExecutor(ClusterService clusterService,
+                           @Nullable BiConsumer<String, IndexItem> constraintsChecker,
+                           @Nullable Runnable validatorsCacheCleaner,
```

Comment:
I think that it's not really nice that we require here on logic which is done outside of this component. But I also have no good suggestion right now how we could add any post-completion logic outside of this component. Maybe @mfussenegger has a better idea?
As least we should change the name to be more general as for the component this is just a runnable run after completion, what it does is not known.
```suggestion
                           Runnable onCompletion,
```

Reply:
I think previous solution with having cache in `ShardedRequests` and **auto** clearance in `close()` had its own flaws but since a new approach not much better because of having post completion logic outside, maybe we can get back to it?